File: Julia/Gemini2.5/NBodySimulationKD.jl
using LinearAlgebra
using Printf
using Random
using Base.Threads
using Statistics # For median

# --- Constants ---
const G = 1.0
const SOFTENING = 0.01
const SOFTENING_SQ = SOFTENING^2
const THETA = 0.3       # Opening angle for Barnes-Hut approximation
const THETA_SQ = THETA^2
const MAX_LEAF_SIZE = 8 # Maximum particles in a leaf node

# --- Data Structures for k-D Tree ---

struct AxisAlignedBoundingBox
    min_coord::Vector{Float64}
    max_coord::Vector{Float64}
end

# Helper function to get the center of a box
center(box::AxisAlignedBoundingBox) = 0.5 .* (box.min_coord .+ box.max_coord)

# Helper function to get the size (max dimension) of a box
box_size(box::AxisAlignedBoundingBox) = maximum(box.max_coord .- box.min_coord)

# Helper function to compute bounding box for a set of particles
function compute_bounds(indices::AbstractVector{Int}, positions::Matrix{Float64})::AxisAlignedBoundingBox
    if isempty(indices)
        # Return an empty/invalid box? Or handle upstream? Let's return zero box.
        return AxisAlignedBoundingBox(zeros(3), zeros(3))
    end
    # Use view for efficiency
    particle_positions = view(positions, indices, :)
    min_c = vec(minimum(particle_positions, dims=1))
    max_c = vec(maximum(particle_positions, dims=1))
    # Add small epsilon to avoid zero-size boxes if all particles coincide
    epsilon = 1e-9
    max_c .+= ifelse.(min_c .== max_c, epsilon, 0.0)
    return AxisAlignedBoundingBox(min_c, max_c)
end

# Helper function to merge two bounding boxes
function merge_bounds(b1::AxisAlignedBoundingBox, b2::AxisAlignedBoundingBox)::AxisAlignedBoundingBox
    min_c = min.(b1.min_coord, b2.min_coord)
    max_c = max.(b1.max_coord, b2.max_coord)
    return AxisAlignedBoundingBox(min_c, max_c)
end

mutable struct KDTreeNode
    bounds::AxisAlignedBoundingBox
    center_of_mass::Vector{Float64}
    total_mass::Float64
    particle_indices::Union{Vector{Int}, Nothing} # Indices if leaf, nothing otherwise
    children::Union{Vector{KDTreeNode}, Nothing}  # Child nodes if internal, nothing otherwise
    # Debug/Info (Optional)
    axis::Int # Splitting axis (0, 1, or 2) -> 1,2,3 in Julia
    split_value::Float64
    is_leaf::Bool

    # Constructor for internal node
    KDTreeNode(bounds, com, mass, children, axis, split_val) =
        new(bounds, com, mass, nothing, children, axis, split_val, false)

    # Constructor for leaf node
    KDTreeNode(bounds, com, mass, indices) =
        new(bounds, com, mass, indices, nothing, -1, NaN, true)
end


# --- k-D Tree Construction ---

"""
Builds the k-D tree recursively.
"""
function build_kdtree(indices::Vector{Int}, positions::Matrix{Float64}, masses::Vector{Float64}; depth::Int = 0)::KDTreeNode
    num_particles = length(indices)
    current_bounds = compute_bounds(indices, positions)

    # Base case: If few particles, create a leaf node
    if num_particles <= MAX_LEAF_SIZE
        com = zeros(3)
        total_mass = 0.0
        if num_particles > 0
            # Use views for particle data within the leaf
            leaf_positions = view(positions, indices, :)
            leaf_masses = view(masses, indices)
            total_mass = sum(leaf_masses)
            if total_mass > 1e-12 # Avoid division by zero if masses are tiny/zero
                 # Weighted average for center of mass: sum(m_i * r_i) / sum(m_i)
                 com = vec(sum(leaf_positions .* leaf_masses, dims=1) ./ total_mass)
            else # Assign geometric center if total mass is negligible
                 com = center(current_bounds)
            end
        end
        # Ensure indices are copied if needed downstream
        return KDTreeNode(current_bounds, com, total_mass, copy(indices))
    end

    # Recursive step: Choose axis and split
    axis = (depth % 3) + 1 # Cycle through axes 1, 2, 3 (x, y, z)

    # Find median coordinate along the axis
    # Using quickselect (via partialsort!) is faster O(N) than full sort O(N log N)
    coords = view(positions, indices, axis)
    median_idx = div(num_particles, 2) + 1 # Index for median value
    # partialsort! rearranges 'indices' such that the median element's original index
    # is placed correctly if indices were sorted by positions[:, axis].
    # It partitions 'indices' based on the values in 'coords'.
    median_split_value = partialsort!(copy(coords), median_idx) # Value used for split
    
    # Partition indices based on the median value
    left_indices = filter(idx -> positions[idx, axis] < median_split_value, indices)
    right_indices = filter(idx -> positions[idx, axis] >= median_split_value, indices)

    # Handle cases where median split doesn't partition well (e.g., many identical coords)
     if isempty(left_indices) || isempty(right_indices)
         # Fallback: simple split in the middle, might not be balanced
         mid = div(num_particles, 2)
         sorted_indices = sort(indices, by=idx -> positions[idx, axis])
         left_indices = sorted_indices[1:mid]
         right_indices = sorted_indices[mid+1:end]
         # Recalculate median_split_value if needed for node info, though less critical now
         if !isempty(right_indices)
            median_split_value = positions[right_indices[1], axis]
         end
     end

    # Build children recursively
    left_child = build_kdtree(left_indices, positions, masses, depth=depth+1)
    right_child = build_kdtree(right_indices, positions, masses, depth=depth+1)

    # Create internal node
    children = [left_child, right_child]
    merged_bounds = merge_bounds(left_child.bounds, right_child.bounds)
    total_mass = left_child.total_mass + right_child.total_mass
    com = zeros(3)
    if total_mass > 1e-12
         com = (left_child.center_of_mass .* left_child.total_mass .+
                right_child.center_of_mass .* right_child.total_mass) ./ total_mass
    else
         com = center(merged_bounds)
    end

    return KDTreeNode(merged_bounds, com, total_mass, children, axis, median_split_value)
end


# --- Force Calculation using k-D Tree ---

"""
Recursive helper function to traverse the tree and calculate acceleration
for a target particle. Accumulates acceleration directly into `accel_accum`.
"""
function force_traverse_recursive!(accel_accum::Vector{Float64}, # Pass accumulator by reference
                                   target_idx::Int,
                                   target_pos::SubArray{Float64}, # Use view for target position
                                   node::KDTreeNode,
                                   positions::Matrix{Float64},
                                   masses::Vector{Float64})

    # Vector from target particle to node's center of mass
    dx = node.center_of_mass[1] - target_pos[1]
    dy = node.center_of_mass[2] - target_pos[2]
    dz = node.center_of_mass[3] - target_pos[3]

    dist_sq = dx^2 + dy^2 + dz^2

    # Avoid self-interaction if target is the *only* particle in a node
    # (More robust check needed if leaf nodes can contain the target particle)
    # This check is implicitly handled by leaf node iteration later.
    # Need check to avoid zero distance if target is exactly at node COM
    if dist_sq < 1e-18 # Effectively zero distance
        return # Skip interaction
    end

    # Barnes-Hut Criterion: s^2 / d^2 < theta^2
    node_size = box_size(node.bounds)
    if (node_size^2 / dist_sq < THETA_SQ) || node.is_leaf
        # If node is far enough OR it's a leaf node:

        if node.is_leaf
            # Iterate through particles in the leaf
            for leaf_idx in node.particle_indices
                if leaf_idx == target_idx
                    continue # Skip self-interaction
                end
                # Direct calculation for particles in the leaf
                dx_leaf = positions[leaf_idx, 1] - target_pos[1]
                dy_leaf = positions[leaf_idx, 2] - target_pos[2]
                dz_leaf = positions[leaf_idx, 3] - target_pos[3]
                dist_sq_leaf = dx_leaf^2 + dy_leaf^2 + dz_leaf^2 + SOFTENING_SQ
                inv_dist_cubed_leaf = G / (dist_sq_leaf * sqrt(dist_sq_leaf))
                force_factor_leaf = masses[leaf_idx] * inv_dist_cubed_leaf

                accel_accum[1] += force_factor_leaf * dx_leaf
                accel_accum[2] += force_factor_leaf * dy_leaf
                accel_accum[3] += force_factor_leaf * dz_leaf
            end
        else
            # Treat internal node as a single pseudo-particle (if far enough)
            dist_sq_soft = dist_sq + SOFTENING_SQ
            inv_dist_cubed = G / (dist_sq_soft * sqrt(dist_sq_soft))
            force_factor = node.total_mass * inv_dist_cubed

            accel_accum[1] += force_factor * dx
            accel_accum[2] += force_factor * dy
            accel_accum[3] += force_factor * dz
        end
    else
        # Node is too close, traverse children (if internal node)
        if !node.is_leaf
            for child_node in node.children
                # Only traverse children that might contain interactions (optional optimization: check bounds intersection)
                # Basic version: always traverse children if node is too close
                 force_traverse_recursive!(accel_accum, target_idx, target_pos, child_node, positions, masses)
            end
        # If it's a leaf but somehow failed the s/d < theta check (shouldn't happen with || node.is_leaf condition),
        # the leaf particle interactions would be handled here too, but the logic above covers it.
        end
    end
end


"""
Calculates accelerations for all bodies using the k-D tree (Barnes-Hut).
Multithreaded loop over particles, each traversing the tree.
"""
function calculate_accelerations_kdtree!(accelerations::Matrix{Float64},
                                         positions::Matrix{Float64},
                                         masses::Vector{Float64})
    N = size(positions, 1)
    particle_indices = collect(1:N)

    # --- Build Tree (Sequential) ---
    # println("Building k-D tree...") # Optional debug message
    # build_time = @elapsed root_node = build_kdtree(particle_indices, positions, masses)
    root_node = build_kdtree(particle_indices, positions, masses)
    # println("Tree build time: $build_time seconds") # Optional debug message

    # --- Calculate Forces (Parallel using Tree) ---
    fill!(accelerations, 0.0) # Reset accelerations

    Threads.@threads for i in 1:N
        # Each thread calculates acceleration for its assigned particle 'i'
        # Need thread-local accumulator or direct write to accelerations[i,:]
        # Direct write is safe as only one thread writes to accelerations[i,:]
        accel_i = zeros(3) # Thread-local accumulator vector
        target_pos_view = view(positions, i, :) # Use view for efficiency
        force_traverse_recursive!(accel_i, i, target_pos_view, root_node, positions, masses)
        accelerations[i, :] = accel_i
    end
end


# --- Simulation Step ---
"""
Performs one step using Kick-Step, calling the k-D tree acceleration function.
"""
function kick_step_update!(positions::Matrix{Float64},
                          velocities::Matrix{Float64},
                          masses::Vector{Float64},
                          accelerations::Matrix{Float64},
                          dt::Float64)

    # Calculate accelerations using k-D Tree approximation
    calculate_accelerations_kdtree!(accelerations, positions, masses)

    # 1. Kick: Update velocities (v_new = v_old + a * dt)
    velocities .+= accelerations .* dt

    # 2. Step: Update positions (p_new = p_old + v_new * dt)
    positions .+= velocities .* dt
end


# --- Energy Calculation (Exact O(N^2) - FOR REFERENCE ONLY with Tree Code) ---
# NOTE: This calculates the EXACT potential energy. The tree code uses approximate forces,
# so the energy calculated by THIS function will NOT be conserved by the simulation dynamics.
# It serves as a baseline reference.

function calculate_total_energy(positions::Matrix{Float64},
                                velocities::Matrix{Float64},
                                masses::Vector{Float64})::Float64
    N = size(positions, 1)
    kinetic_energy = 0.0

    # Kinetic Energy (O(N)) - Sequential
    for i in 1:N
        vel_sq = velocities[i, 1]^2 + velocities[i, 2]^2 + velocities[i, 3]^2
        kinetic_energy += 0.5 * masses[i] * vel_sq
    end

    # Exact Potential Energy (O(N^2)) - Parallelized
    num_threads = Threads.nthreads()
    pe_partials = zeros(Float64, num_threads)

    Threads.@threads for i in 1:N
        tid = Threads.threadid()
        pe_thread = 0.0
        for j in (i+1):N
            dx = positions[j, 1] - positions[i, 1]
            dy = positions[j, 2] - positions[i, 2]
            dz = positions[j, 3] - positions[i, 3]
            dist_sq = dx^2 + dy^2 + dz^2 + SOFTENING_SQ
            dist = sqrt(dist_sq)
            pe_thread -= G * masses[i] * masses[j] / dist
        end
        pe_partials[tid] += pe_thread
    end
    potential_energy = sum(pe_partials)

    return kinetic_energy + potential_energy
end


# --- Initialization --- (No changes needed here)
function initialize_circular_orbits(N_orbiting::Int,
                                    central_mass::Float64;
                                    min_radius::Float64 = 1.0,
                                    max_radius::Float64 = 10.0,
                                    orbiting_mass_factor::Float64 = 1e-7
                                    )::Tuple{Matrix{Float64}, Matrix{Float64}, Vector{Float64}}
    N_total = N_orbiting + 1
    positions = zeros(Float64, N_total, 3)
    velocities = zeros(Float64, N_total, 3)
    masses = zeros(Float64, N_total)
    masses[1] = central_mass
    rng = MersenneTwister()
    for i in 2:N_total
        masses[i] = central_mass * orbiting_mass_factor * (0.5 + rand(rng))
        radius = min_radius + (max_radius - min_radius) * rand(rng)
        phi = 2.0 * pi * rand(rng)
        cos_theta = 2.0 * rand(rng) - 1.0
        sin_theta = sqrt(max(0.0, 1.0 - cos_theta^2))
        x = radius * sin_theta * cos(phi)
        y = radius * sin_theta * sin(phi)
        z = radius * cos_theta
        positions[i, :] = [x, y, z]
        vel_magnitude = sqrt(G * central_mass / radius)
        pos_vec = positions[i, :]
        if abs(pos_vec[1]) < 1e-9 && abs(pos_vec[2]) < 1e-9
             axis_vec = [1.0, 0.0, 0.0]
        else
             axis_vec = [0.0, 0.0, 1.0]
        end
        vel_dir = cross(pos_vec, axis_vec)
        if norm(vel_dir) < 1e-9
             axis_vec = [0.0, 1.0, 0.0]
             vel_dir = cross(pos_vec, axis_vec)
        end
        norm_vel_dir = norm(vel_dir)
        if norm_vel_dir > 1e-9
            vel_dir_normalized = vel_dir / norm_vel_dir
        else
            vel_dir_normalized = normalize([rand(rng)-0.5, rand(rng)-0.5, rand(rng)-0.5])
            # println("Warning: Could not determine unique velocity direction for body $i. Using random.")
        end
        velocities[i, :] = vel_magnitude .* vel_dir_normalized
    end
    return positions, velocities, masses
end

# --- Main Simulation Function ---
function run_simulation(num_bodies_orbiting::Int, num_steps::Int, dt::Float64)
    num_threads = Threads.nthreads()
    if num_threads == 1
        println("WARNING: Julia started with only 1 thread. Multithreading will be limited.")
        println("         Restart Julia with '-t auto' or '-t N' (e.g., julia -t 4 script.jl)")
    end
    println("--- N-Body Simulation Start (k-D Tree, Threads: $num_threads) ---")
    println("Algorithm: Barnes-Hut approximation with k-D Tree")
    println("Opening Angle (theta): ", THETA)
    println("Number of orbiting bodies: ", num_bodies_orbiting)
    println("Total bodies: ", num_bodies_orbiting + 1)
    println("Number of steps: ", num_steps)
    println("Time step (dt): ", dt)
    println("---------------------------------")

    println("Initializing system...")
    central_mass = 1.0e6
    positions, velocities, masses = initialize_circular_orbits(num_bodies_orbiting, central_mass)
    N_total = size(positions, 1)
    accelerations = zeros(Float64, N_total, 3)
    println("Initialization complete.")

    println("Calculating initial exact energy (Reference)...")
    initial_energy = calculate_total_energy(positions, velocities, masses) # Exact O(N^2) calc
    @printf("Initial Exact Energy: %.6e\n", initial_energy)
    println("---------------------------------")

    println("Starting simulation loop...")
    start_time = time()

    for step in 1:num_steps
        # Store positions before step for energy calc if needed, or calculate E less often
        # pos_before = copy(positions) # Example if needed

        kick_step_update!(positions, velocities, masses, accelerations, dt) # Uses k-D Tree

        if step % max(1, num_steps ÷ 10) == 0 || step == num_steps
           elapsed_time = time() - start_time
           # Optional: Calculate exact energy periodically to see drift
           # current_exact_energy = calculate_total_energy(positions, velocities, masses)
           # @printf("Step: %d / %d (%.1f%%), Elapsed: %.2f s, Exact Energy: %.6e\n",
           #         step, num_steps, 100 * step / num_steps, elapsed_time, current_exact_energy)
           @printf("Step: %d / %d (%.1f%%), Elapsed Time: %.2f s\n",
                   step, num_steps, 100 * step / num_steps, elapsed_time)
        end
    end

    end_time = time()
    total_time = end_time - start_time
    println("Simulation loop finished.")
    println("---------------------------------")

    println("Calculating final exact energy (Reference)...")
    final_energy = calculate_total_energy(positions, velocities, masses) # Exact O(N^2) calc
    @printf("Final Exact Energy:   %.6e\n", final_energy)

    energy_diff = abs(final_energy - initial_energy)
    relative_error = if abs(initial_energy) > 1e-12
        energy_diff / abs(initial_energy)
    else
        energy_diff
    end
    println("** NOTE: Energy calculated using exact O(N^2) method. **")
    println("** Simulation used approximate O(N log N) tree forces. **")
    println("** Therefore, this energy difference reflects the **")
    println("** approximation error, not just numerical integration error. **")
    @printf("Absolute Exact Energy Difference: %.6e\n", energy_diff)
    @printf("Relative Exact Energy Error: %.6e (%.4f%%)\n", relative_error, relative_error * 100)
    println("---------------------------------")
    @printf("Total Simulation Time: %.3f seconds\n", total_time)
    println("--- N-Body Simulation End ---")
end

# --- Run ---
# Remember to start Julia with threads: julia -t auto nbody_kdtree.jl

# Now N=1,000,000 might be feasible (though still demanding) due to O(N log N)
const NUM_ORBITING_BODIES = 100_000 # Reduced from 1M for faster testing
# const NUM_ORBITING_BODIES = 1_000_000 # Uncomment for full scale (will take time!)
const NUM_STEPS = 100 # Reduced steps for faster testing
# const NUM_STEPS = 1000 # Uncomment for full run
const DT = 0.001

if isinteractive() && NUM_ORBITING_BODIES > 50000
    println("Running k-D Tree simulation with N_orbiting = $(NUM_ORBITING_BODIES) for $(NUM_STEPS) steps using $(Threads.nthreads()) threads.")
    println("Theta = $THETA. This may still take significant time and memory.")
    println("Press Enter to continue, or Ctrl+C to abort.")
    readline()
elseif !isinteractive() && Threads.nthreads() == 1
     println("WARNING: Running script with only 1 thread. Performance will be limited.")
     println("         Consider restarting Julia with the -t flag for multithreading.")
end

run_simulation(NUM_ORBITING_BODIES, NUM_STEPS, DT)

File: Python/Gemini2.5/NBodySimulationKD.py
import numpy as np
import time
import multiprocessing as mp
from functools import partial

# --- Global variables for worker processes ---
# Using globals is a common pattern with multiprocessing to avoid data serialization overhead.
g_pos = None
g_mass = None
g_kdtree_root = None
g_theta = 0.0

class KDNode:
    """A node in the k-D Tree for Barnes-Hut simulation."""
    def __init__(self, particle_indices, bbox):
        self.particle_indices = np.array(particle_indices, dtype=int)
        self.bbox = bbox
        self.is_leaf = len(self.particle_indices) <= 1
        
        self.left_child = None
        self.right_child = None
        
        self.total_mass = 0.0
        self.center_of_mass = np.zeros(3)
        self.size = np.max(self.bbox[1] - self.bbox[0])

def build_kdtree(particle_indices, bbox, depth=0):
    """Recursively builds the k-D tree."""
    node = KDNode(particle_indices, bbox)
    if node.is_leaf:
        return node

    # Determine split dimension and value
    split_dim = depth % 3
    particle_positions = g_pos[particle_indices]
    
    # Find the median to split the particles into two halves
    median_idx = np.argsort(particle_positions[:, split_dim])[len(particle_indices) // 2]
    median_particle_index = particle_indices[median_idx]
    split_val = g_pos[median_particle_index, split_dim]

    left_indices = particle_indices[particle_positions[:, split_dim] < split_val]
    right_indices = particle_indices[particle_positions[:, split_dim] >= split_val]

    # Create new bounding boxes for children
    left_bbox = np.copy(bbox)
    left_bbox[1, split_dim] = split_val
    node.left_child = build_kdtree(left_indices, left_bbox, depth + 1)

    right_bbox = np.copy(bbox)
    right_bbox[0, split_dim] = split_val
    node.right_child = build_kdtree(right_indices, right_bbox, depth + 1)
    
    return node

def annotate_tree(node):
    """Recursively calculates the center of mass and total mass for each node."""
    if node.is_leaf:
        if len(node.particle_indices) > 0:
            idx = node.particle_indices[0]
            node.total_mass = g_mass[idx]
            node.center_of_mass = g_pos[idx]
    else:
        annotate_tree(node.left_child)
        annotate_tree(node.right_child)
        
        m1 = node.left_child.total_mass
        m2 = node.right_child.total_mass
        node.total_mass = m1 + m2
        
        if node.total_mass > 0:
            node.center_of_mass = (m1 * node.left_child.center_of_mass + 
                                  m2 * node.right_child.center_of_mass) / node.total_mass
    return node

def calculate_force_on_particle(particle_idx, node):
    """Calculates the gravitational force on a single particle by traversing the tree."""
    accel = np.zeros(3)
    
    # Use a stack for non-recursive traversal
    stack = [node]
    while stack:
        current_node = stack.pop()
        
        # Do not interact with yourself
        if len(current_node.particle_indices) == 1 and current_node.particle_indices[0] == particle_idx:
            continue
            
        # Calculate distance and size
        r_vec = current_node.center_of_mass - g_pos[particle_idx]
        dist_sq = np.sum(r_vec**2)
        dist = np.sqrt(dist_sq)

        if dist > 0:
            # Barnes-Hut criterion: s/d < theta
            if current_node.size / dist < g_theta or current_node.is_leaf:
                # Treat as a macro-particle
                accel += current_node.total_mass * r_vec / (dist_sq * dist)
            else:
                # Node is too close, traverse its children
                if current_node.right_child: stack.append(current_node.right_child)
                if current_node.left_child: stack.append(current_node.left_child)
    return accel

def worker_calculate_force_chunk(particle_indices):
    """Worker function for multiprocessing. Calculates force for a chunk of particles."""
    chunk_accel = np.zeros((len(particle_indices), 3))
    for i, idx in enumerate(particle_indices):
        chunk_accel[i] = calculate_force_on_particle(idx, g_kdtree_root)
    return chunk_accel
    
def calculate_accelerations_parallel(pool, n_chunks):
    """Orchestrates parallel force calculation."""
    all_indices = np.arange(len(g_mass))
    index_chunks = np.array_split(all_indices, n_chunks)
    
    accel_results = pool.map(worker_calculate_force_chunk, index_chunks)
    
    # Stitch results back together
    return np.vstack(accel_results)

def init_worker(mass, pos, vel, root, theta):
    """Initializer for each worker process."""
    global g_mass, g_pos, g_vel, g_kdtree_root, g_theta
    g_mass, g_pos, g_vel = mass, pos, vel
    g_kdtree_root, g_theta = root, theta

# (The initialization and energy calculation functions from the previous example
#  can be reused here without modification. For brevity, they are omitted but
#  should be included in the final script.)
def initialize_circular_orbits(n_bodies, central_mass_val):
    total_bodies = n_bodies + 1; mass = np.ones(total_bodies); pos = np.zeros((total_bodies, 3)); vel = np.zeros((total_bodies, 3)); mass[0] = central_mass_val
    for i in range(1, total_bodies):
        radius = np.random.uniform(5.0, 150.0); theta_ang = np.arccos(2 * np.random.random() - 1); phi = 2 * np.pi * np.random.random()
        pos[i] = [radius * np.sin(theta_ang) * np.cos(phi), radius * np.sin(theta_ang) * np.sin(phi), radius * np.cos(theta_ang)]
        orbital_speed = np.sqrt(central_mass_val / radius); random_vec = np.random.randn(3)
        while np.allclose(np.cross(pos[i], random_vec), 0): random_vec = np.random.randn(3)
        vel[i] = np.cross(pos[i], random_vec); vel[i] = vel[i] / np.linalg.norm(vel[i]) * orbital_speed
    return mass, pos, vel
def calculate_energy(mass, pos, vel):
    ke = 0.5 * np.sum(mass[:, np.newaxis] * vel**2); pe = 0
    for i in range(pos.shape[0]):
        for j in range(i + 1, pos.shape[0]):
            r_vec = pos[j] - pos[i]; r = np.linalg.norm(r_vec)
            if r > 0: pe -= (mass[i] * mass[j]) / r
    return ke, pe
    
def main():
    # Simulation parameters
    # NOTE: 1M bodies is very memory intensive for k-D trees. 
    # A number like 100,000 is more feasible on consumer hardware.
    n_orbiting_bodies = 100_000 
    central_body_mass = 1_000_000
    n_steps = 10
    dt = 0.01
    theta = 0.3

    # Parallelization parameters
    num_processes = mp.cpu_count()
    n_chunks = num_processes * 4

    print("Initializing system...")
    mass, pos, vel = initialize_circular_orbits(n_orbiting_bodies, central_body_mass)
    
    # Calculate initial energy using direct summation for accuracy
    print("Calculating initial energy (direct summation)...")
    ke_initial, pe_initial = calculate_energy(mass, pos, vel)
    total_energy_initial = ke_initial + pe_initial
    print(f"Initial Total Energy: {total_energy_initial:.5e}")

    print(f"\nStarting simulation on {num_processes} cores with theta = {theta}...")
    sim_start_time = time.time()
    
    # Main simulation loop
    for step in range(n_steps):
        # 1. Build the k-D Tree (sequentially)
        min_coords = np.min(pos, axis=0)
        max_coords = np.max(pos, axis=0)
        bbox = np.array([min_coords, max_coords])
        root_node = build_kdtree(np.arange(len(mass)), bbox)

        # 2. Annotate the Tree with CoM info (sequentially)
        # We update the global position reference for the annotation function
        global g_pos, g_mass
        g_pos = pos
        g_mass = mass
        annotate_tree(root_node)
        
        # 3. Calculate forces in parallel using the tree
        with mp.Pool(processes=num_processes, initializer=init_worker, initargs=(mass, pos, vel, root_node, theta)) as pool:
            accel = calculate_accelerations_parallel(pool, n_chunks)

        # 4. Update velocities and positions (Kick-Step)
        vel += accel * dt
        pos += vel * dt

        if (step + 1) % 100 == 0:
            print(f"  - Step {step + 1}/{n_steps} completed.")
    
    sim_end_time = time.time()
    print(f"Simulation loop took: {sim_end_time - sim_start_time:.2f} seconds")

    print("\nCalculating final energy (direct summation)...")
    ke_final, pe_final = calculate_energy(mass, pos, vel)
    total_energy_final = ke_final + pe_final
    print(f"Final Total Energy:   {total_energy_final:.5e}")
    
    energy_change = abs((total_energy_final - total_energy_initial) / total_energy_initial)
    print(f"\nRelative change in total energy: {energy_change:.5%}")


if __name__ == "__main__":
    main()